### PR TITLE
Add auto-upgrade feature via configMap update

### LINKED ISSUE #5: Add Compatibility Matrix Support for VDO
**Issue/Feature Description:**
Add the compatibility matrix support for the VDO(vSphereKubernetes Driver Operator) for the deployment of CSI and CPI Drivers via operator.
As a part of this feature request we intend to find the compatible versions of CSI and CPI with k8s and vSphere and deploy the driver accordingly. It will also help in auto-upgrade of the drivers.

**What type of Issue/Feature is this?**
 /kind new feature

**Why this fix/ feature is needed(give scenarios or use cases):**
This feature is required to ensure that VDO installs/upgrade the correct versions of driver each time and also give flexibility to Users to choose their own versions and related deployment files.

### PR DESCRIPTION
**Pull request by asifdxtreme**
**What type of PR is this?**
> /kind enhancement

**What this PR does / why we need it**:
This PR adds the support of upgrade scenario where user can update the versions of CPI/CSI without restarting the vdo-controller. 
User need to update the configMap which contains the upgraded version Matrix, and VDO will dynamically listen to any change in the configMap and apply the changes accordingly.

**Which issue(s) this PR fixes**:
Fixes #5 

**Test Report Added?**:
> /kind TESTED

**Special notes for your reviewer**:
 - [x] addition of UT in Progress

### CODE REVIEW

**Review by lgayatri — APPROVED** [review]

File: main.go [hunk 1]
```diff
@@ -93,12 +93,17 @@ func main() {
...
+	vdoConfigReconciler := &controllers.VDOConfigReconciler{
 		Client:       mgr.GetClient(),
 		Logger:       ctrllog.Log.WithName("controllers").WithName("VDOConfig"),
 		Scheme:       mgr.GetScheme(),
 		ClientConfig: mgr.GetConfig(),
-	}).SetupWithManager(mgr); err != nil {
+	}
+
+	//Start configMap watcher for Compatibility Matrix
+	go vdoConfigReconciler.WatchForConfigMapChanges()
```
Comment (jvrahav) [inline main.go:104]: why not just setup another watch when setting up the manager 
```

			Watches(
				&source.Kind{Type: &v1.ConfigMap{}},
				handler.EnqueueRequestsFromMapFunc(func(object client.Object) []reconcile.Request {
					cm, _ := object.(*v1.ConfigMap)
					if cm.Namespace == VDO_NAMESPACE && cm.Name == "test-cm" {
						
					}
					return nil
				}))
```
we expect there to be only vdoconfig resource in the cluster. Hence here in the watch we can fetch the vdo config resource and raise a reconcile request for the resource.
we should be leveraging the constructs in controller-runtime libraries to achieve these functionalities
we should also always compare the actual vs desired and perform reconcile to achieve the desired function

**Review by jvrahav — CHANGES_REQUESTED** [review]

File: controllers/vdoconfig_controller.go [hunk 2]
```diff
@@ -787,6 +839,20 @@ func (r *VDOConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
 					}
 				}
 
+				return nil
+			})).
+		Watches(
+			&source.Kind{Type: &v1.ConfigMap{}},
+			handler.EnqueueRequestsFromMapFunc(func(object client.Object) []reconcile.Request {
+				configMap, _ := object.(*v1.ConfigMap)
```
Comment (jvrahav) [inline controllers/vdoconfig_controller.go:847]: what happens if the configmap is deleted. should we unset the environment variables

**Review by jvrahav — CHANGES_REQUESTED** [review]

File: controllers/vdoconfig_controller_test.go [hunk 3]
```diff
@@ -1103,6 +1103,147 @@ var _ = Describe("TestGetMatrixConfig", func() {
...
+				Datacenters:    nil,
+				VsphereVersion: "7.0.3",
+			}, nil
+
+		}
+		vdoConfig := initializeVDOConfig()
+		Expect(r.Create(vdoctx, vdoConfig)).Should(Succeed())
+
+		//Setup VC SIM
+		model := simulator.VPX()
```
Comment (jvrahav) [inline controllers/vdoconfig_controller_test.go:1168]: we can avoid spawning a vc sim instance here as we are only looking to create a configmap with vc creds, which can be any random text. bringing up vc sim for this purpose seems to be an excess

**Review by jvrahav — APPROVED** [review]
can we add a copy of the matrix.yaml in the code base.

**Pull request merged**